### PR TITLE
07 db

### PR DESCRIPTION
**Pull request by ZagorskayaMarina**
Add DB functionality
Add splitting classes into packages

### CODE REVIEW

File: OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
+package com.marina.dao.DBConnection;
```
Comment (artemmoroz) [inline OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java]: wrong package name

File: OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.sql.*;
+import java.util.Properties;
+
+public class DatabaseConnection {
+    private static DatabaseConnection uniqueInstance;
+    private Connection connection;
+
+    private static String path = "c:/Users/user/Java/AT_ISsoft/AT_ISsoft/OnlineStore/dao/src/main/resources/db.properties";
```
Comment (artemmoroz) [inline OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java]: u can not use any absolute path in your apps

File: OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    private DatabaseConnection(){
+        try {
+            this.connection = getDataSource(prop).getConnection();
+        } catch (SQLException e) {
+            throw new RuntimeException("Something wrong with data source", e);
+        }
+    };
+
+    public static DatabaseConnection getInstance() throws SQLException {
+        if (uniqueInstance == null) {
```
Comment (artemmoroz) [inline OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java]: you need smthing for synchronization your instance

File: OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/CategoriesDAO.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        Connection connection = null;
+        Statement statement = null;
+        int result;
+        try {
+            connection = DatabaseConnection.getInstance().getConnection();
+            statement = connection.createStatement();
+            result = statement.executeUpdate(query);
+            System.out.println(result);
+        } catch (SQLException throwables) {
+            throwables.printStackTrace();
```
Comment (artemmoroz) [inline OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/CategoriesDAO.java]: now a very good option just to print stacktrace

File: OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/CategoriesDAO.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CategoriesDAO {
+
+    CommonDAO commonDAO = new CommonDAO();
```
Comment (artemmoroz) [inline OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/CategoriesDAO.java]: inheritance?

File: OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/ProductsDAO.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ProductsDAO {
+
+    CommonDAO commonDAO = new CommonDAO();
+
+    public void updateProductTable(String query) {
+        Connection connection = null;
```
Comment (artemmoroz) [inline OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/ProductsDAO.java]: very similar class to categoryDAO

File: OnlineStore/domain/src/main/java/com/marina/entities/Product.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class Product {
+    private int productID;
```
Comment (artemmoroz) [inline OnlineStore/domain/src/main/java/com/marina/entities/Product.java]: you should not merge domain model and db model classes

File: OnlineStore/domain/src/main/java/com/marina/entities/Product.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    private String name;
+    private int rating;
+    private double price;
+    private int categoryID;
```
Comment (artemmoroz) [inline OnlineStore/domain/src/main/java/com/marina/entities/Product.java]: and also naming conventions (camelCase)

File: OnlineStore/store/src/main/java/com/marina/utility/Utils.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+public class Utils {
+
+    public static int generateRandomValue(int min, int max){
+        return (int)(min + Math.random() * max);
+    }
+
+    public static int getIntFromConsole(InputStream inputStream){
+        Scanner in = new Scanner(inputStream);
+        System.out.println("Input product id: ");
```
Comment (artemmoroz) [inline OnlineStore/store/src/main/java/com/marina/utility/Utils.java:14]: text in service method (imp)

File: OnlineStore/store/src/main/java/com/marina/utility/Utils.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
+package com.marina.utility;
+
+import java.io.InputStream;
+import java.util.Scanner;
+
+public class Utils {
+
+    public static int generateRandomValue(int min, int max){
+        return (int)(min + Math.random() * max);
```
Comment (artemmoroz) [inline OnlineStore/store/src/main/java/com/marina/utility/Utils.java:9]: you can use faker for that (imp)

File: OnlineStore/store/src/main/java/com/marina/services/impl/BasicProductPublisher.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import com.marina.dao.commonDAO.ProductsDAO;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class BasicProductPublisher {
+    private Faker faker = new Faker();
+
+    protected List<Product> publicProducts(List<Category> categories){
+        List<Product> products = new ArrayList<>();
```
Comment (artemmoroz) [inline OnlineStore/store/src/main/java/com/marina/services/impl/BasicProductPublisher.java]: discuss

File: OnlineStore/domain/src/main/java/com/marina/entities/Product.java (hunk 7 above)
Comment (artemmoroz) [reply OnlineStore/domain/src/main/java/com/marina/entities/Product.java]: lib: bean copy utils, objects merge

File: OnlineStore/store/src/main/java/com/marina/services/impl/ReflectionPublisher.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+
+    public List<Category> getCategoriesNameViaReflection() throws ClassNotFoundException, IllegalAccessException, InstantiationException, NoSuchFieldException {
+        List<Category> categories= new ArrayList<>();
+        Reflections reflections = new Reflections("com.marina.categories");
+        Set<Class<? extends Category>> subClassesForCategory = reflections.getSubTypesOf(Category.class);
+        for (Class clazz : subClassesForCategory) {
+            Object o = Class.forName(clazz.getName()).newInstance();
+            int categoryID = clazz.getField("categoryID").getInt(o);
```
Comment (artemmoroz) [inline OnlineStore/store/src/main/java/com/marina/services/impl/ReflectionPublisher.java]: Category o = (Category) Class.forName(clazz.getName()).newInstance();
        categories.add(o);

**Review by artemmoroz — CHANGES_REQUESTED** [review]

File: OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java (hunk 1 above)
Comment (ZagorskayaMarina) [reply OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java]: fixed

File: OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java (hunk 2 above)
Comment (ZagorskayaMarina) [reply OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java]: fixed

File: OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java (hunk 3 above)
Comment (ZagorskayaMarina) [reply OnlineStore/dao/src/main/java/com/marina/dao/DBConnection/DatabaseConnection.java]: fixed

File: OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/CategoriesDAO.java (hunk 5 above)
Comment (ZagorskayaMarina) [reply OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/CategoriesDAO.java]: fixed

File: OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/ProductsDAO.java (hunk 6 above)
Comment (ZagorskayaMarina) [reply OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/ProductsDAO.java]: fixed

File: OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/CategoriesDAO.java (hunk 4 above)
Comment (ZagorskayaMarina) [reply OnlineStore/dao/src/main/java/com/marina/dao/commonDAO/CategoriesDAO.java]: fixed

File: OnlineStore/domain/src/main/java/com/marina/entities/Product.java (hunk 7 above)
Comment (ZagorskayaMarina) [reply OnlineStore/domain/src/main/java/com/marina/entities/Product.java]: fixed

File: OnlineStore/domain/src/main/java/com/marina/entities/Product.java (hunk 8 above)
Comment (ZagorskayaMarina) [reply OnlineStore/domain/src/main/java/com/marina/entities/Product.java]: fixed

File: OnlineStore/store/src/main/java/com/marina/services/impl/BasicProductPublisher.java (hunk 11 above)
Comment (ZagorskayaMarina) [reply OnlineStore/store/src/main/java/com/marina/services/impl/BasicProductPublisher.java]: impl

File: OnlineStore/store/src/main/java/com/marina/services/impl/ReflectionPublisher.java (hunk 12 above)
Comment (ZagorskayaMarina) [reply OnlineStore/store/src/main/java/com/marina/services/impl/ReflectionPublisher.java]: fixed

File: OnlineStore/data/src/main/java/com/marina/dao/CategoriesDAO.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+
+public class CategoriesDAO extends CommonDAO<Category> {
+
+    @Override
+    public void completeStatement(PreparedStatement statement, List<Category> list) {
```
Comment (artemmoroz) [inline OnlineStore/data/src/main/java/com/marina/dao/CategoriesDAO.java:13]: it is better to thow ex (e/g unssoported operation, or similar) or implement method. Just not leave him empty)

**Review by artemmoroz — APPROVED** [review]

**Pull request merged**